### PR TITLE
8318480: Obsolete UseCounterDecay and remove CounterDecayMinIntervalLength

### PR DESCRIPTION
**Pull request by dlunde**
This changeset obsoletes the leftover (i.e., no longer used for anything) product compiler flag `UseCounterDecay` (requires CSR) and removes the leftover develop flag `CounterDecayMinIntervalLength`.

Changes:
- Obsolete `UseCounterDecay` in JDK 22 and expire it in JDK 23.
- Completely remove `CounterDecayMinIntervalLength`.

### Testing
Platforms: windows-x64, linux-x64, linux-aarch64, macosx-x64, macosx-aarch64.
- `tier1`
- HotSpot parts of `tier2` and `tier3`

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Change requires CSR request [JDK-8320162](https://bugs.openjdk.org/browse/JDK-8320162) to be approved
- [x] Commit message must refer to an issue

### Issues
 * [JDK-8318480](https://bugs.openjdk.org/browse/JDK-8318480): Obsolete UseCounterDecay and remove CounterDecayMinIntervalLength (**Enhancement** - P4)
 * [JDK-8320162](https://bugs.openjdk.org/browse/JDK-8320162): Obsolete UseCounterDecay (**CSR**)

### Reviewers
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**)
 * [Leonid Mesnik](https://openjdk.org/census#lmesnik) (@lmesnik - **Reviewer**) ⚠️ Review applies to [cd3343fe](https://git.openjdk.org/jdk/pull/16673/files/cd3343fee068b09f51a14fcc224b147088fa02d2)
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16673/head:pull/16673` \
`$ git checkout pull/16673`

Update a local copy of the PR: \
`$ git checkout pull/16673` \
`$ git pull https://git.openjdk.org/jdk.git pull/16673/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16673`

View PR using the GUI difftool: \
`$ git pr show -t 16673`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16673.diff">https://git.openjdk.org/jdk/pull/16673.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16673#issuecomment-1812123623)

### CODE REVIEW

**Review by TobiHartmann — COMMENTED** [review]
I think the description of `UseCounterDecay` in globals.hpp should be changed to `"Adjust recompilation counters (deprecated)"`

**Comment by dlunde** [comment]
> I think the description of `UseCounterDecay` in globals.hpp should be changed to `"Adjust recompilation counters (deprecated)"`

Thanks, updated.

**Review by TobiHartmann — APPROVED** [review]
Looks good to me.

**Review by lmesnik — APPROVED** [review]
Please update copyrgiths.

**Comment by dlunde** [comment]
> Please update copyrgiths.

Right, thanks @lmesnik. Updated.

**Review by dholmes-ora — COMMENTED** [review]
As `UseCounterDecay` already does nothing there is no need to have a deprecation step - just go straight to obsoletion. The only difference to the user will be the message they see.

**Comment by dlunde** [comment]
Thanks @dholmes-ora, I agree (but was under the impression you always had to deprecate flags as a first step). I'll revise the CSR and changeset to directly obsolete the flag instead.

**Review by dholmes-ora — APPROVED** [review]
LGTM! Thanks

**Review by TobiHartmann — APPROVED** [review]
Still good.

File: src/hotspot/share/runtime/globals.hpp [hunk 1]
```diff
@@ -1220,16 +1220,9 @@ const int ObjectAlignmentInBytes = 8;
   product(bool, UseCompiler, true,                                          \
           "Use Just-In-Time compilation")                                   \
                                                                             \
-  product(bool, UseCounterDecay, true,                                      \
-          "Adjust recompilation counters")                                  \
-                                                                            \
   develop(intx, CounterHalfLifeTime,    30,                                 \
           "Half-life time of invocation counters (in seconds)")             \
                                                                             \
```
Comment (ysramakrishna) [inline src/hotspot/share/runtime/globals.hpp:1225]: Thanks for making these changes. However, it seems like the obsolete `CounterHalfLifeTime` was missed in this cleanup?

Comment (dlunde) [reply src/hotspot/share/runtime/globals.hpp:1225]: @ysramakrishna Thanks and yes, it should also be removed. I'll make a new issue and PR.

Comment (dlunde) [reply src/hotspot/share/runtime/globals.hpp:1225]: For reference: https://github.com/openjdk/jdk/pull/21034

**Pull request closed without merging**